File: 2024/Day_03/part02.py
import re
import numpy as np

with open('test.txt', 'r') as file:
    mem = file.read()

clean_mem = ''
i = 0
do_block = True 

# iterates through memory string to only include 'do' parts
while i < len(mem):
    if do_block:
        if mem[i:i+5] == "don't":
            do_block = False 
            i += 5
        else:
            clean_mem+=mem[i]
            i += 1
    else:
        if mem[i:i+2] == 'do' and mem[i:i+5] != "don't":
            do_block = True 
            i += 2
        else:
            i += 1


pattern = r"mul\((\d{1,3}),(\d{1,3})\)"

matches = np.array(re.findall(pattern, clean_mem), dtype=int)
total = np.sum(np.prod(matches, axis=1))
print(total)

File: 2024/Day_03/part02.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <numeric>
#include <sstream>
#include <regex>
#include <string>
using namespace std;

void readFile(string &memoryData) {
    ifstream file("test.txt");
    if (!file.is_open()){
        cerr << "Error Opening File" << endl;
        exit(-1);
    }

    char ch;
    while (file.get(ch)) memoryData += ch;
    file.close();
}

string cleanData(string &memoryData) {
    string cleanMemory = "";
    int i = 0;
    bool do_block = true;

    while (i < memoryData.length()) {
        if (do_block) {
            if (memoryData.substr(i, 5) == "don't") {
                do_block = false;
                i += 5;
            }
            else {
                cleanMemory += memoryData[i];
                i++;
            }
        } else {
            if (memoryData.substr(i, 2) == "do" && memoryData.substr(i, 5) != "don't") {
                do_block = true;
                i += 2;
            } else {
                i++;
            }
        }
    }
    return cleanMemory;
}

int countPattern(string &memoryData, regex pattern) {
    vector<pair<int, int>> matches;

    // finds matches
    sregex_iterator begin(memoryData.begin(), memoryData.end(), pattern);
    sregex_iterator end;    

    // collects matches
    for (sregex_iterator it = begin; it != end; it++) {
        int num1 = stoi((*it)[1].str());
        int num2 = stoi((*it)[2].str());
        matches.push_back({num1, num2});
    }

    int numOccur = 0;
    for (vector<pair<int, int>>::iterator it = matches.begin(); it != matches.end(); it++) {
        numOccur += it->first * it->second;
    }
    return numOccur;
}

int main() {
    string memoryData;
    regex pattern(R"(mul\((\d{1,3}),(\d{1,3})\))");
    int numOccur = 0;

    readFile(memoryData);
    string cleanMemory = cleanData(memoryData);
    numOccur = countPattern(cleanMemory, pattern);

    cout << numOccur << endl;
    return 0;
}